compiler/src/lnntoamm/Types.ts

```diff
@@ -197,83 +252,426 @@ class Struct extends Type {
     return this; // TODO: this right?
   }
 
-  tempConstrain(to: Type) {
+  tempConstrain(to: Type, scope: Scope) {
     // TODO: can structs have temp constraints?
-    this.constrain(to);
+    this.constrain(to, scope);
   }
 
   resetTemp() {
     // TODO: can structs have temp constraints?
   }
+
+  dupIfNotLocalInterface(): Type | null {
+    return null;
+  }
+}
+
+abstract class Has extends Type {
+  constructor(
+    name: string,
+    ast: LPNode | null,
+  ) {
+    super(name, ast);
+  }
+
+  static field(field: HasField, ty: Type): boolean {
+    // TODO: structs
+    return false;
+  }
+
+  static method(method: HasMethod, scope: Scope, ty: Type): Fn[] {
+    let fns = scope.get(method.name);
+    if (!isFnArray(fns)) {
+      return [];
+    }
+    return Fn.select(fns, method.params.map(p => p === null ? ty : p), scope);
+  }
+
+  static operator(operator: HasOperator, scope: Scope, ty: Type): Operator[] {
+    let ops: Operator[] = scope.get(operator.name);
+    // if there is no op by that name, RIP
+    if (!isOpArray(ops)) {
+      return [];
+    }
+    // filter out ops that aren't the same fixity
+    ops = ops.filter(op => op.isPrefix === operator.isPrefix);
+    if (operator.isPrefix) {
+      return ops.filter(op => op.select(scope, operator.params[0] || ty) !== []);
+    } else {
+      return ops.filter(op => op.select(scope, operator.params[0] || ty, operator.params[1] || ty) !== []);
+	  }
```

Comment:
Nit: accidental tab character here, I think.

```suggestion
    }
```